zipline/data/data_portal.py

```diff
@@ -696,6 +766,25 @@ def get_history_window(self, assets, end_dt, bar_count, frequency, field,
 
         # forward-fill if needed
         if field == "price" and ffill:
+            assets_with_nan_index = np.where(pd.isnull(df.iloc[0, :]))[0]
+            assets_with_leading_nan = [
```

Comment:
Something like this?

```
diff --git a/zipline/data/data_portal.py b/zipline/data/data_portal.py
index 8383f46..ddf449d 100644
--- a/zipline/data/data_portal.py
+++ b/zipline/data/data_portal.py
@@ -752,8 +752,9 @@ class DataPortal(object):
             raise ValueError("Invalid history field: " + str(field))

         # sanity check in case sids were passed in
-        assets = [(self.env.asset_finder.retrieve_asset(asset) if
-                   isinstance(asset, int) else asset) for asset in assets]
+        assets = np.array([
+            (self.env.asset_finder.retrieve_asset(asset) if
+             isinstance(asset, int) else asset) for asset in assets])

         if frequency == "1d":
             df = self._get_history_daily_window(assets, end_dt, bar_count,
@@ -767,8 +768,7 @@ class DataPortal(object):
         # forward-fill if needed
         if field == "price" and ffill:
             assets_with_nan_index = np.where(pd.isnull(df.iloc[0, :]))[0]
-            assets_with_leading_nan = [
-                assets[i] for i in assets_with_nan_index]
+            assets_with_leading_nan = assets[assets_with_nan_index]

             if frequency == "1m":
                 data_frequency = 'minute'
```